File: 1101-1200/1125-Smallest Sufficient Team/1125-Smallest Sufficient Team.cpp
class Solution {
public:
    vector<int> smallestSufficientTeam(vector<string>& req_skills, vector<vector<string>>& people) {
        unordered_map<string, int> table;
        for (int i = 0; i < req_skills.size(); ++i) {
            table[req_skills[i]] = (1 << i);
        }
        
        int n = req_skills.size();
        unordered_map<int, vector<int>> dp;
        dp.reserve(1 << n); // use reserved space to avoid rehashing, otherwise iterrator update in dp can cause wrong result 
        dp[0] = {};
        for (int i = 0; i < people.size(); ++i) {
            int curr = 0;
            for (string s : people[i]) {
                if (table.count(s)) {
                    curr |= table[s];
                }
            }
            
            for (auto p : dp) {
                int newSkills = p.first | curr;
                if (newSkills != p.first) {
                    if (!dp.count(newSkills) || dp[newSkills].size() > p.second.size() + 1) {
                        dp[newSkills] = p.second;
                        dp[newSkills].push_back(i);
                    }
                }
            }
        }

        return dp[(1 << n) - 1];
    }
};

File: 1101-1200/1125-Smallest Sufficient Team/1125-Smallest Sufficient Team.py
class Solution:
    def smallestSufficientTeam(self, req_skills: List[str], people: List[List[str]]) -> List[int]:
        table = {skill : (1 << i) for i, skill in enumerate(req_skills)}
        dp = {0 : []}
        for i, p in enumerate(people):
            curr = 0
            for s in p:
                if s in table:
                    curr |= table[s]

            for skills, team in list(dp.items()):
                newSkills = skills | curr
                if newSkills != skills:
                    if newSkills not in dp or len(dp[newSkills]) > len(team) + 1:
                        dp[newSkills] = team + [i]
        
        return dp[(1 << len(req_skills)) - 1]
